pyemu/en.py

```diff
@@ -811,15 +811,72 @@ def phi_vector(self):
         """
         cols = self._df.columns
         pst = self.pst
-        weights = self.pst.observation_data.loc[cols, "weight"]
-        obsval = self.pst.observation_data.loc[cols, "obsval"]
+        weights = pst.observation_data.loc[cols, "weight"]
+        obsval = pst.observation_data.loc[cols, "obsval"]
         phi_vec = []
         for idx in self._df.index.values:
             simval = self._df.loc[idx, cols]
             phi = (((simval - obsval) * weights) ** 2).sum()
             phi_vec.append(phi)
         return pd.Series(data=phi_vec, index=self.index)
 
+    def get_phi_vector_noise_obs(self, noise_obs_filename=None):
+        """vector of L2 norm (phi) for the realizations (rows) of `Ensemble`. This
+            differs from `phi_vector` in that phi is calculated for each ensemble 
+            member with respect to the realization of observation noise for that 
+            ensemble member represented in the `noise_obs_filename` argument.
+
+        Args:
+            noise_obs_filename (_type_, string): Filename of csv or binary noise realizations
+                                                ensemble filename. If None, attemps to use the 
+                                                filename from the `pst` attribute of `self`.
+                                                Defaults to None.
+        Returns:
+            `pandas.Series`: series of realization name (`Ensemble.index`) and phi values
+        
+        """
+        # if no noise_obs_filename included, try to grab from pst object
+        if noise_obs_filename is None:
+            try:
+                noise_obs_filename = self.pst.pestpp_options['ies_observation_ensemble']
+            except:
+                raise Exception(
+                        "no noise_observation_filename passed or found in pst options"
+                    )
+        # if it looks like a binary file, try loading from binary
+        if (noise_obs_filename.endswith('jcb')) or (noise_obs_filename.endswith('jco')):
+            try:
+                noise_obs = pyemu.Matrix.from_binary(noise_obs_filename).to_dataframe()
+            except:
+                raise Exception(
+                    f"Could not open {noise_obs_filename}"
+                )
+        # otherwise assume it's a csv
+        else:
+            kwargs = {}
+            kwargs["index_col"] = 0
+            kwargs["low_memory"] = False
+            try:
+                noise_obs = pd.read_csv(noise_obs_filename, **kwargs)
+            except:
+                raise Exception(
+                    f"Could not open {noise_obs_filename}"
+                )
+        # make sure the index is all string (since the real)
+        noise_obs.index = [str(i) for i in noise_obs.index]
+        
+        # now make the calculations
+        cols = self._df.columns
+        weights = self.pst.observation_data.loc[cols, "weight"]
+        phi_vec = []
+        for idx in self._df.index.values:
```

Comment:
wonder if you could get away with:
`phi_vec = (((self._df-noise_obs) * weights)**2).dropna().sum(axis=1)`
pandas should align nicely on index and columns right?

Reply:
maybe, but @jtwhite79  pointed out that we need to handle inequality constraints, so gotta sum by groups and handle conditions about violation of inequality constraints. This led to some refactoring which may be close to done. If we get it working, I still like the idea of de-looping but not sure we can

Reply:
ah yea, good call. I think should be poss to work with a residual ensemble `res = (self._df-noise_obs)` and then zero out the residual that are the wrong sign for obs that are in inequality groups. `res.loc[:, ltobs] = res.loc[:, ltobs].clip(0)`

Reply:
that sounds like a good idea. I might try and look tomorrow. If you get a chance, maybe check out the latest commit I pushed. I think it's working, but too slow to be practical. It worked for a simple case without noise, but the noise obs example is a bigger file and def too slow. Late here but I can look tomorrow or Friday again

Reply:
All the tests are passing now so I think we are good, but the ensemble version is pretty slow. I like your idea @briochh so might try and implement that way